src/state/slices/stakingDataSlice/selectors.ts

```diff
@@ -169,14 +221,122 @@ export const selectRewardsByAccountSpecifier = createDeepEqualOutputSelector(
   }
 )
 
+export const selectAllRewardsByDenom = createDeepEqualOutputSelector(
+  selectStakingDataByAccountSpecifier,
+  selectDenom,
+  (stakingData, denom): Record<string, chainAdapters.cosmos.Reward[]> => {
+    if (!stakingData || !stakingData.rewards) return {}
+
+    const rewards = stakingData.rewards.reduce(
+      (acc: Record<string, chainAdapters.cosmos.Reward[]>, current: ValidatorReward) => {
+        if (!acc[current.validator.address]) {
+          acc[current.validator.address] = []
+        }
+
+        acc[current.validator.address].push(
+          ...current.rewards.filter(x => ASSET_ID_TO_DENOM[x.assetId] === denom)
+        )
+
+        return acc
+      },
+      {}
+    )
+
+    return rewards
+  }
+)
+
 export const selectRewardsAmountByDenom = createSelector(
   selectRewardsByAccountSpecifier,
+  selectValidatorAddress,
   selectDenom,
-  (rewardsByAccountSpecifier, denom): string => {
+  (rewardsByAccountSpecifier, validatorAddress, denom): string => {
     if (!rewardsByAccountSpecifier.length) return ''
 
     const rewards = rewardsByAccountSpecifier.find(rewards => rewards.denom === denom)
 
     return rewards?.amount || ''
   }
 )
+
+export const selectAllValidators = createDeepEqualOutputSelector(
+  selectStakingData,
+  stakingData => stakingData.byvalidator
+)
+
+export const selectSingleValidator = createSelector(
+  selectStakingData,
+  selectValidatorAddress,
+  (stakingData, validatorAddress) => {
+    return stakingData.byvalidator[validatorAddress] || null
+  }
+)
+
+export const selectNonloadedValidators = createSelector(
+  selectStakingDataByAccountSpecifier,
+  selectAllValidators,
+  (stakingData, allValidators): string[] => {
+    if (!stakingData) return []
+    const initialValidatorsAddresses = stakingData.delegations
+      ? stakingData.delegations.map(x => x.validator.address)
+      : []
+    if (stakingData.undelegations) {
+      initialValidatorsAddresses.push(...stakingData.undelegations.map(x => x.validator.address))
+    }
+
+    if (stakingData.rewards) {
+      initialValidatorsAddresses.push(...stakingData.rewards.map(x => x.validator.address))
+    }
+
+    const uniqueValidatorAddresses = [...new Set(initialValidatorsAddresses)]
+    return uniqueValidatorAddresses.filter(x => allValidators[x] === undefined)
+  }
+)
+
+export const selectStakingOpportunityDataByDenom = createDeepEqualOutputSelector(
```

Comment:
I wonder if we should not have any of the "ByDenom" selectors.

I originally didnt realuze that we never actually pass "denom" to the frontend. Its translated to assetId in chain adapters and I think thats fine. IMO theres no reason web should ever know or care about denom as it is a cosmos specific thing and we really only care about our own asset id format (caip19)

Reply:
Do you think denom can go entirely for now then and we will never have any issues in case of multiple e.g opportunities/rewards/delegations/undelegations in different denoms?

Reply:
If we use caip19 instead of denom and do the translation in the chain-adapter layer I dont think theres any reason for denom in web. There should be no problem supporting multiple assets.

We would do it here:

https://github.com/shapeshift/lib/blob/main/packages/chain-adapters/src/cosmossdk/CosmosSdkBaseAdapter.ts#L95

Currently the frontend has no concept of denom its just assuming the same assetId as the chain adapter. It wouldnt be hard to add a simple translation in chain adapters to convert it all to chainId and then its in the native format that our stack expects.

I dont like how we use ASSET_ID_TO_DENOM to construct an artificial denom to use in web. Its just like we are jumping through hoops for no reason and I think it would be a lot cleaner if web only had knowledge of assetId (caip19)

Reply:
Another advantage of keeping it as `assetId` is when new developers help on the frontend they dont need any cosmos-specific domain knowledge. We can just translate it to assetId very early in the stack (chain-adapters) and I think everything will "just work" and be less confusing

Reply:
I do agree with the rationale, and we could very well do with web not knowing about denom. Since it will be added as a metadata in asset-service most likely, we could make that translation in chain-adapters and let web not know anything about it. Until that is implemented there however, we are stuck either with:
- picking all rewards (also delegations, undelegations, redelegations) and assuming that it is the same denom, with no filtering (that would work for Cosmos). This is effectively what you propose, except for the fact that it's not implemented in chain-adapters of course. 
- having some kind of discrimination to get the data for the right denom. This is more future-proof, and while we hardcode this in `ASSET_ID_TO_DENOM`, it will eventually go away and we'll be able to either leave the front-end denom-agnostic, or use that logic directly from the `Asset` type

I'm happy if we do switch to the first option though. It will need a bit more thinking for e.g Osmosis, which is fine since that's a whole different milestone.

Reply:
Anyone taking over this, I've had some discussion with @elmutt and we'll remove denom logic for now. Let's just assume that all the data we get is the same denom, which actually is in the case of Cosmos. Web should become agnostic about denoms.

Reply:
> Anyone taking over this, I've had some discussion with @elmutt and we'll remove denom logic for now. Let's just assume that all the data we get is the same denom, which actually is in the case of Cosmos. Web should become agnostic about denoms.

I don't think we should assume that they are all the same denom (asset). We will eventually support multiple assets and should keep coding with that assumption.

I just mean we should only be using assetId (caip19) and not denom. caip19 already encompasses all possible denoms and is the standard identifier we use in our stack. Passing denom through to web just makes things more complicated